fri/src/fold_even_odd.rs

```diff
@@ -28,20 +29,93 @@ pub fn fold_even_odd<F: TwoAdicField>(poly: &[F], beta: F) -> Vec<F> {
 
     let n = poly.len();
     debug_assert!(n > 1);
+
     let log_n = log2_strict_usize(n);
 
     let g_inv = F::two_adic_generator(log_n).inverse();
     let one_half = F::TWO.inverse();
     let half_beta = beta * one_half;
 
     // beta/2 times successive powers of g_inv
-    let powers = Powers {
-        base: g_inv,
-        current: half_beta,
-    };
+    let powers = g_inv.packed_powers::<F::Packing>(half_beta);
+
+    // pack first / second polys, rounding up to the nearest multiple of packing width
+    let half_n = n / 2;
+    let cutoff = (half_n / F::Packing::WIDTH) * F::Packing::WIDTH;
 
     let (first, second) = poly.split_at(n / 2);
-    izip!(powers, first, second)
-        .map(|(power, a, b)| (one_half + power) * *a + (one_half - power) * *b)
-        .collect()
+    let first_leftover = F::Packing::from_fn(|i| {
+        if cutoff + i < first.len() {
+            first[i]
+        } else {
+            F::ZERO
+        }
+    });
+    let second_leftover = F::Packing::from_fn(|i| {
+        if cutoff + i < second.len() {
+            second[cutoff + i]
+        } else {
+            F::ZERO
+        }
+    });
+    let first = F::Packing::pack_slice(&first[..cutoff])
+        .iter()
+        .chain(core::iter::once(&first_leftover));
+    let second = F::Packing::pack_slice(&second[..cutoff])
+        .iter()
+        .chain(core::iter::once(&second_leftover));
+
+    // allocate and pack result, rounding up to the nearest multiple of packing width
+    let nearest_mutliple_of_packing_width =
+        F::Packing::WIDTH * ceil_div_usize(half_n, F::Packing::WIDTH);
+    let mut res = vec![F::ZERO; nearest_mutliple_of_packing_width];
+    let res_packed = F::Packing::pack_slice_mut(&mut res);
+
+    let one_half = F::Packing::from_fn(|_| one_half);
+    for (src, dst) in izip!(powers, first, second)
+        .map(|(power, &a, &b)| (one_half + power) * a + (one_half - power) * b)
+        .zip(res_packed.iter_mut())
+    {
+        *dst = src;
+    }
```

Comment:
nits: `.iter_mut()` can be omitted, and maybe slightly easier to read like

```suggestion
    let packed_results = izip!(powers, first, second)
        .map(|(power, &a, &b)| (one_half + power) * a + (one_half - power) * b);
    for (src, dst) in packed_results.zip(F::Packing::pack_slice_mut(&mut res)) {
        *dst = src;
    }
```